tests/src/test/scala/whisk/core/cli/test/ApiGwTests.scala

```diff
@@ -62,137 +61,146 @@ class ApiGwTests
       * If number of CLI invocations in this suite have reached the throttle limit
       * then pause the test for enough time so that the throttle restriction is gone
       */
-    def checkThrottle() = {
-      // If the # CLI calls at the throttle limit, then wait enough time to avoid the CLI being blocked
-      cliCallCount += 1
-      println(s"Action invokes ${cliCallCount}; per minute thottle limit ${maxActionsPerMin}")
-      if ( cliCallCount > (maxActionsPerMin - 10) ) {  // Allow for some margin...10 invocations
-        println(s"Action invokes ${cliCallCount} close to exceeding per minute thottle limit of ${maxActionsPerMin}")
-        val waitedAlready = Duration.fromNanos(java.time.Duration.between(clearedThrottleTime, Instant.now).toNanos)
-        settleThrottle(waitedAlready)
-        cliCallCount = 0
-        clearedThrottleTime = Instant.now
-      }
-    }
-
-    /**
-     * Settles throttles of 1 minute. Waits up to 1 minute depending on the time already waited.
-     *
-     * @param waitedAlready the time already gone after the last invoke or fire
-     */
-    def settleThrottle(waitedAlready: FiniteDuration) = {
-      val timeToWait = (throttleWindow - waitedAlready).max(Duration.Zero)
-      println(s"Waiting for ${timeToWait.toSeconds} seconds, already waited for ${waitedAlready.toSeconds} seconds")
-      Thread.sleep(timeToWait.toMillis)
+    def checkThrottle(maxInvocationsBeforeThrottle: Int = maxActionsPerMin) = {
+        var invocationCount = 0
+        var t = Instant.now
+        var tminus60 = t.minusSeconds(60)
+        var throttleTime = 0L
+        for (invocationTime <- invocationTimes) {
+            if (invocationTime.isAfter(tminus60)) {
+                // Find the "oldest" invocation time during this 60 second window
+                if (throttleTime == 0L) {
+                    // Instead of waiting a fixed N seconds to settle the throttle,
+                    // calulate a wait time that will clear out about half of the
+                    // current invocations (assuming even distribution) from the
+                    // next 60 second period.  This is guaranteed to clear out at
+                    // least one invocation without having to wait the full 60 seconds.
+                    val millisToOldInvocation = t.toEpochMilli - invocationTime.toEpochMilli
+                    throttleTime = 60.seconds.toMillis - (millisToOldInvocation / 2)
+                }
+                invocationCount += 1
+            }
+        }
```

Comment:
This looks like you want: Get the first invocation after a certain  `Instant` and get a count of all invocations after a certain `Instant`, is that correct?

If so, you might want to think about Scala's collection methods, like:

```
val relevantInvocations = invocationTimes.filter(_.isAfter(tminus60))
val throttleTime = relevantInvocations.headOption.map { invocationTime =>
    val millisToOldInvocation = t.toEpochMilli - invocationTime.toEpochMilli
    60.seconds.toMillis - (millisToOldInvocation / 2)
}.getOrElse(0)

val invocationCount = relevantInvocations.size
```